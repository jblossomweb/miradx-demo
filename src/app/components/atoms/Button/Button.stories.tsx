import React from 'react'
import { BiUserPlus as UserPlus } from 'react-icons/bi'
import type { Meta, StoryObj } from '@storybook/react'

import Spinner from '../Spinner'
import Button from './Button'

type Story = StoryObj<typeof meta>

const meta = {
  title: 'components/atoms/Button',
  component: Button,
  render: (args) => <Button {...args} />,
} satisfies Meta<typeof Button>

export const Sample: Story = {
  args: {
    label: 'Sample Button',
  },
}

export const Slate: Story = {
  args: {
    label: 'Slate Button',
    color: 'slate',
  },
}

export const Teal: Story = {
  args: {
    label: 'Teal Button',
    color: 'teal',
  },
}

export const Disabled: Story = {
  args: {
    label: 'Disabled Button',
    disabled: true,
  },
}

export const Icon: Story = {
  args: {
    label: 'Sample Button',
    icon: <UserPlus size={24} />,
  },
}

export const DisabledIcon: Story = {
  args: {
    label: 'Disabled Button',
    icon: <Spinner size={24} />,
    disabled: true,
  },
}

export default meta
