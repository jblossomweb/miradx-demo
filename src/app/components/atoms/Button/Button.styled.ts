import tw from 'tailwind-styled-components'
import type { AllowedColors } from './Button'

export interface ButtonProps {
  $color: AllowedColors
}

const bgClass = (color: AllowedColors) => {
  switch (color) {
    case 'slate':
      return 'bg-slate-500 hover:bg-slate-700'
    case 'teal':
      return 'bg-teal-600 hover:bg-teal-800'
  }
}

export const Button = tw.button<ButtonProps>`
  font-bold
  py-2 px-4
  rounded-full
  flex
  items-center
  ${({ disabled, $color }) =>
    disabled
      ? 'cursor-not-allowed text-gray-400 bg-gray-500'
      : `cursor-pointer text-white ${bgClass($color)}`}
`

export const Icon = tw.span`
  ml-1
`
