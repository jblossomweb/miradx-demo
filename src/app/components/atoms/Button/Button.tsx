import React from 'react'
import * as Styled from './Button.styled'

export type AllowedColors = 'teal' | 'slate'

export interface ButtonProps {
  label?: string
  type?: 'submit' | 'reset' | 'button'
  color?: AllowedColors
  disabled?: boolean
  icon?: React.ReactNode
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button: React.FC<ButtonProps> = ({
  label,
  type = 'button',
  color = 'teal',
  disabled,
  icon,
  onClick,
}) => (
  <Styled.Button
    type={type}
    disabled={disabled}
    onClick={onClick}
    $color={color}
  >
    {label}
    {!!icon && <Styled.Icon>{icon}</Styled.Icon>}
  </Styled.Button>
)

export default Button
