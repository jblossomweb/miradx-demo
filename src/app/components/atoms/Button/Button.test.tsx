import React from 'react'
import { render, screen } from '@testing-library/react'
import storySnaps from 'test/utils/storySnaps'

import Button from './Button'
import * as stories from './Button.stories'

describe('app/components/atoms/Button', () => {
  describe('stories', () => {
    storySnaps(stories)
  })

  describe('assertions', () => {
    const onClick = jest.fn()
    const renderDefault = () => render(<Button />)
    const renderDisabled = () => render(<Button disabled />)
    const renderHandler = () => render(<Button onClick={onClick} />)

    beforeEach(jest.clearAllMocks)

    it('renders a button', () => {
      renderDefault()
      const button = screen.getByRole('button') as HTMLButtonElement
      expect(button).toBeInTheDocument()
    })

    it('renders a button as not disabled by default', () => {
      renderDefault()
      const button = screen.getByRole('button') as HTMLButtonElement
      expect(button.disabled).not.toBe(true)
    })

    it('renders a disabled button when specified', () => {
      renderDisabled()
      const button = screen.getByRole('button') as HTMLButtonElement
      expect(button.disabled).toBe(true)
    })

    it('calls onClick when button is clicked', () => {
      renderHandler()
      const button = screen.getByRole('button') as HTMLButtonElement
      expect(onClick).not.toHaveBeenCalled()
      button.click()
      expect(onClick).toHaveBeenCalled()
    })
  })
})
