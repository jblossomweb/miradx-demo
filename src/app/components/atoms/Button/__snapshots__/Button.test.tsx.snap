// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/atoms/Button stories Disabled matches markup snapshot 1`] = `
<DocumentFragment>
  <button
    class="font-bold py-2 px-4 rounded-full flex items-center cursor-not-allowed text-gray-400 bg-gray-500"
    disabled=""
    type="button"
  >
    Disabled Button
  </button>
</DocumentFragment>
`;

exports[`app/components/atoms/Button stories DisabledIcon matches markup snapshot 1`] = `
<DocumentFragment>
  <button
    class="font-bold py-2 px-4 rounded-full flex items-center cursor-not-allowed text-gray-400 bg-gray-500"
    disabled=""
    type="button"
  >
    Disabled Button
    <span
      class="ml-1"
    >
      <svg
        class="animate-spinfast"
        fill="currentColor"
        height="24"
        role="status"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 22c5.421 0 10-4.579 10-10h-2c0 4.337-3.663 8-8 8s-8-3.663-8-8c0-4.336 3.663-8 8-8V2C6.579 2 2 6.58 2 12c0 5.421 4.579 10 10 10z"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`app/components/atoms/Button stories Icon matches markup snapshot 1`] = `
<DocumentFragment>
  <button
    class="font-bold py-2 px-4 rounded-full flex items-center cursor-pointer text-white bg-teal-600 hover:bg-teal-800"
    type="button"
  >
    Sample Button
    <span
      class="ml-1"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 8h-2v3h-3v2h3v3h2v-3h3v-2h-3zM4 8a3.91 3.91 0 0 0 4 4 3.91 3.91 0 0 0 4-4 3.91 3.91 0 0 0-4-4 3.91 3.91 0 0 0-4 4zm6 0a1.91 1.91 0 0 1-2 2 1.91 1.91 0 0 1-2-2 1.91 1.91 0 0 1 2-2 1.91 1.91 0 0 1 2 2zM4 18a3 3 0 0 1 3-3h2a3 3 0 0 1 3 3v1h2v-1a5 5 0 0 0-5-5H7a5 5 0 0 0-5 5v1h2z"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`app/components/atoms/Button stories Sample matches markup snapshot 1`] = `
<DocumentFragment>
  <button
    class="font-bold py-2 px-4 rounded-full flex items-center cursor-pointer text-white bg-teal-600 hover:bg-teal-800"
    type="button"
  >
    Sample Button
  </button>
</DocumentFragment>
`;

exports[`app/components/atoms/Button stories Slate matches markup snapshot 1`] = `
<DocumentFragment>
  <button
    class="font-bold py-2 px-4 rounded-full flex items-center cursor-pointer text-white bg-slate-500 hover:bg-slate-700"
    type="button"
  >
    Slate Button
  </button>
</DocumentFragment>
`;

exports[`app/components/atoms/Button stories Teal matches markup snapshot 1`] = `
<DocumentFragment>
  <button
    class="font-bold py-2 px-4 rounded-full flex items-center cursor-pointer text-white bg-teal-600 hover:bg-teal-800"
    type="button"
  >
    Teal Button
  </button>
</DocumentFragment>
`;
